realm/realm-library/src/main/java/io/realm/RealmTypeSelectorForMap.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * Copyright 2020 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.realm;
+
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import javax.annotation.Nullable;
+
+import io.realm.internal.OsMap;
+import io.realm.internal.OsResults;
+import io.realm.internal.RealmObjectProxy;
+import io.realm.internal.Table;
+import io.realm.internal.util.Pair;
+
+/**
+ * Abstracts certain operations from value operators depending on the type of Realm we are working
+ * with.
+ *
+ * @param <K> the key type
+ * @param <V> the value type
+ */
+abstract class RealmTypeSelectorForMap<K, V> {
+
+    protected final BaseRealm baseRealm;
+    protected final OsMap osMap;
+
+    RealmTypeSelectorForMap(BaseRealm baseRealm, OsMap osMap) {
+        this.baseRealm = baseRealm;
+        this.osMap = osMap;
+    }
+
+    protected V getRealmModel(BaseRealm baseRealm, long realmModelKey) {
+        // do nothing here, override only for RealModels
+        return null;
```

Comment:
Cannot see if it is the case, but consider throwing if this should never be triggered for non-RealmModes to catch unintended calls. 

Similarly below.